### PR TITLE
catalog: Simplify catalog fencing

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, when syncing with new updates, the persist catalog
implemented fencing by looking at both the persist shard upper and the
Epoch. This commit updates that fencing logic to only look at the Epoch
while guaranteeing that all writers increment the epoch. This
simplifies the persist implementation slightly and will allow further
refactors to combine the pre and post open catalog implementations.

### Motivation
This PR refactors existing code.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
@ParkMyCar don't review this yet. I still have some things to work out.

**Comment by jkosh44** [comment]
> @ParkMyCar don't review this yet. I still have some things to work out.

Ok, this is ready now.

**Comment by jkosh44** [comment]
> @ParkMyCar don't review this yet. I still have some things to work out.

Ok, this is ready now.

> > @ParkMyCar don't review this yet. I still have some things to work out.
> 
> Ok, this is ready now.

Ugh, sorry for the churn I just realized something else. I'm putting this back in draft.

**Comment by ParkMyCar** [comment]
> Ugh, sorry for the churn I just realized something else. I'm putting this back in draft.

No problem at all!

**Comment by jkosh44** [comment]
> > @ParkMyCar don't review this yet. I still have some things to work out.
> 
> Ok, this is ready now.
> 
> > > @ParkMyCar don't review this yet. I still have some things to work out.
> > 
> > 
> > Ok, this is ready now.
> 
> Ugh, sorry for the churn I just realized something else. I'm putting this back in draft.

Ok, now it's really ready for review.

**Comment by jkosh44** [comment]
> I retriggered the parallel-workload nightly that went OoM: https://buildkite.com/materialize/nightlies/builds/6987#018e5de4-77de-4e71-8998-73d2e378b328 (in persist-txn, so suspicious)

Looks like it passed on a retry. I'm not sure how this change could result in an OOM, unless the OOM is really masking a different issue, so I think it's unrelated.

File: src/catalog/src/durable/persist.rs [hunk 1]
```diff
@@ -922,22 +934,14 @@ impl PersistCatalogState {
...
-                return Err(DurableCatalogError::Fence(format!(
-                    "current catalog upper {:?} fenced by new catalog upper {:?}",
-                    self.upper, upper
-                ))
-                .into());
-            }
-        }
-
+        self.sync(upper).await?;
+        self.epoch.validate()?;
```
Comment (ParkMyCar) [inline src/catalog/src/durable/persist.rs:944]: Looks like previously we would sync even if there was an upper mismatch, but if we were in a read_only mode (e.g. catalog-debug?). Did we lose that behavior?

**Review by ParkMyCar — APPROVED** [review]
Makes sense to me!

File: src/catalog/src/durable/persist.rs (hunk 1 above)
Comment (jkosh44) [reply src/catalog/src/durable/persist.rs:944]: I don't think I fully understand the question. Now we always `sync`, even if there is an upper mismatch. This includes read_only mode.

**Pull request merged**